src/coreclr/vm/jithelpers.cpp

```diff
@@ -5366,17 +5374,59 @@ NOINLINE static void JIT_ReversePInvokeEnterRare(ReversePInvokeFrame* frame)
 
     frame->currentThread = thread;
 
+#ifdef PROFILING_SUPPORTED
+        if (CORProfilerTrackTransitions())
+        {
+            ProfilerUnmanagedToManagedTransitionMD(frame->pMD, COR_PRF_TRANSITION_CALL);
+        }
+#endif
+
     thread->DisablePreemptiveGC();
+#ifdef DEBUGGING_SUPPORTED
+    // If the debugger is attached, we use this opportunity to see if
+    // we're disabling preemptive GC on the way into the runtime from
+    // unmanaged code. We end up here because
+    // Increment/DecrementTraceCallCount() will bump
+    // g_TrapReturningThreads for us.
+    if (CORDebuggerTraceCall())
+        g_pDebugInterface->TraceCall((const BYTE*)traceAddr);
+#endif // DEBUGGING_SUPPORTED
 }
 
-NOINLINE static void JIT_ReversePInvokeEnterRare2(ReversePInvokeFrame* frame)
+NOINLINE static void JIT_ReversePInvokeEnterRare2(ReversePInvokeFrame* frame, void* traceAddr)
 {
     frame->currentThread->RareDisablePreemptiveGC();
-}
+#ifdef DEBUGGING_SUPPORTED
+    // If the debugger is attached, we use this opportunity to see if
+    // we're disabling preemptive GC on the way into the runtime from
+    // unmanaged code. We end up here because
+    // Increment/DecrementTraceCallCount() will bump
+    // g_TrapReturningThreads for us.
+    if (CORDebuggerTraceCall())
+        g_pDebugInterface->TraceCall((const BYTE*)traceAddr);
+#endif // DEBUGGING_SUPPORTED
+}
+
+// The following two methods are special.
+// They handle setting up Reverse P/Invoke calls and transitioning back to unmanaged code.
+// As a result, we may not have a thread in JIT_ReversePInvokeEnter and we will be in the wrong GC mode for the HCALL prolog.
+// Additionally, we set up and tear down SEH handlers when we're on x86, so we can't use dynamic contracts anyway.
+// As a result, we specially decorate this method to have the correct calling convention
+// and argument ordering for an HCALL, but we don't use the HCALL macros and contracts
+// since this method doesn't follow the contracts.
+void F_CALL_CONV HCCALL3(JIT_ReversePInvokeEnter, ReversePInvokeFrame* frame, CORINFO_METHOD_HANDLE handle, void* secretArg)
```

Comment:
You have not addressed my comment about additional performance overhead of passing the extra method handle and secret arg arguments. It is introducing performance regression, it is breaking change for R2R (the helper has different signature now), and this design does not work for NativeAOT.

We should be paying this overhead only when somebody turns on the specific profiling notifications (they is rare and slow by design):
- Add a flag to JIT/EE interface that controls whether 
- Keep the existing helper signature as is, introduce a new helper that takes these arguments
- Call the new helper from JIT/EE interface only when the new flag is introduced

Could you please fix that?

Reply:
Opened https://github.com/dotnet/runtime/issues/47223

Reply:
I will take a look at this ASAP.